MessageSigning/MessageSigning.md

```diff
@@ -0,0 +1,437 @@
+Status: Draft
+Type: Standards
+Created: 2020-09-28
+License: CC-BY-4.0
+---
+
+## Abstract
+
+Private keys can be used to sign arbitrary data. If you have the public key, you can verify the data was signed by the owner of the private key. This is how transaction signing works internally but its utility is not limited to transactions. This document tries to set a standard for how to represent and verify signed messages for Cardano.
+
+## Motivation
+
+Most common use cases:
+
+1) Proving ownership of a set addresses (possibly to prove ownership of more then X Ada)
+1) Proving ownership of addresses used in a transaction
+1) Proving ownership of an identity or other off-chain data with a public key attached to it
+
+## Specification
+
+### Requirements
+
+On top being usable for all cases mentioned in *Motivation*, we also desire the following to ensure it works well with hardware wallets:
+
+- E1 - Low runtime memory environment
+- E2 - Low app size environment (cannot implement every cryptographic algorithm on the device or app size would be too big)
+- E3 - Works well with limited display (some hardware wallets cannot display long text and cannot display UTF8)
+
+### Related specifications
+
+#### Requirement Levels [[RFC 2119](https://tools.ietf.org/html/rfc2119)]
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://tools.ietf.org/html/rfc2119).
+
+#### Concise Binary Object Representation (CBOR) [[RFC 7049](https://tools.ietf.org/html/rfc7049)]
+
+CBOR is a way to serialize structured data in a more compact way than what is allowed by JSON. It is widely used across the Cardano ecosystem and so we use it to encode the data for this specification.
+
+#### *Concise Data Definition Language (CDDL)* [[RFC 8610](https://tools.ietf.org/html/rfc8610)]
+
+CDDL is a human-readable CBOR notation format. CBOR schemas defined in this document are defined uinsg CDDL. We use `label = int / tstr` in several places.
+
+#### *CBOR Object Signing and Encryption (COSE)* [[RFC 8152](https://tools.ietf.org/html/rfc8152)] 
+
+This is a standard for how to use CBOR for message signing. It is based on the JSON equivalent *JSON Object Signing and Encryption* [RFC 7520](https://tools.ietf.org/html/rfc7520). 
+
+We base our construction on COSE because all Cardano libraries already depend on CBOR due to its use in the base protocol (which means we don't need to introduce a new library). It is also more compact, which is useful in case data generated by this standard ever needs to be stored on-chain.
+
+Note that we do NOT follow the COSE specification exactly for the following reason:
+
+- We need to restrict the number of algorithms used to only ones that would already be required to provide base Cardano functionality (see E1).
```

Comment:
I'd appreciate it if in the following text it was very clearly indicated which parts come from the COSE spec and which deviate. I'd be tempted to just omit most of the content that exactly follows COSE, perhaps.

Reply:
Using a subset of COSE would be fine I think, but it has to be _implementable_ from pure COSE. So that any library implementing COSE could be used to implement this spec.